### PR TITLE
Implemented smoother selector bt node

### PR DESCRIPTION
**Pull request by owen7900**
Signed-off-by: Owen Hooper <17ofh@queensu.ca>

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3270  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation of Tally |

---

## Description of documentation updates required from your changes
* Added a smoother selector node that is a copy of the controller selector bt node
* Added tests for above selector node

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_behavior_tree/plugins/action/smoother_selector_node.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+{
+  last_selected_smoother_ = msg->data;
+}
+
+}  // namespace nav2_behavior_tree
+
+#include "behaviortree_cpp_v3/bt_factory.h"
+BT_REGISTER_NODES(factory)
+{
+  factory.registerNodeType<nav2_behavior_tree::SmootherSelector>("SmootherSelector");
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/action/smoother_selector_node.cpp:90]: Add this to the BT plugins list in the BT Navigator + in the yaml parameters

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/smoother_selector_node.hpp [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+{
+
+/**
+ * @brief The SmootherSelector behavior is used to switch the smoother
+ * that will be used by the smoother server. It subscribes to a topic "smoother_selector"
+ * to get the decision about what smoother must be used. It is usually used before of
+ * the FollowPath. The selected_smoother output port is passed to smoother_id
+ * input port of the FollowPath
+ */
+class SmootherSelector : public BT::SyncActionNode
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/smoother_selector_node.hpp:38]: Make sure to open a PR for documentation updates: migration guide mentioning this exists, add to the list of navigation plugins, and a configuration guide page for the parameters

**Review by SteveMacenski — APPROVED** [review]
LGTM!

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/smoother_selector_node.hpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2020 Pablo Iñigo Blasco
+//
```
Comment (jwallace42) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/smoother_selector_node.hpp:3]: Add yourself :).

**Comment by jwallace42** [comment]
Would it be worth (maybe future pr) to create a abstract selector node? We currently 
planner_selector, controller_selector, smoother_selector which largely use the same code.

**Comment by SteveMacenski** [comment]
Yeah, we could do that. I haven't worried about it since its a very small amount of code that would be saved, but that's totally sensible.

**Comment by SteveMacenski** [comment]
@owen7900 any update on the documentation PR? :smile:

**Comment by owen7900** [comment]
@SteveMacenski  this week got unexpectedly busy. I'll submit that PR tomorrow morning.

**Comment by SteveMacenski** [comment]
@owen7900 I know we we're talking in a thread somewhere (email? slack? linkedin?  I can't seem to find it now) but would you be interested in another project? You did this perfectly, thanks much!

This would be a nice project as well: https://github.com/ros-planning/navigation2/issues/3150 a little less self-contained but I think the discussion in the issue gives some clear design options and problem description if you wanted something with a little more meat!

**Comment by owen7900** [comment]
@SteveMacenski, I don't think we've been talking anywhere (I'm not sure the slack link is still working, it wouldn't let me use my pre-existing email), but I would happily take that on. If you want to reach out further I would be happy to communicate over slack if I had access. I'll update early next week with progress as Saturday and Monday are my "extra-curricular" days.

**Comment by SteveMacenski** [comment]
Try this: https://join.slack.com/t/navigation2/shared_invite/zt-uj428p0x-jKx8U7OzK1IOWp5TnDS2rA

Oops sorry :laughing:

**Pull request merged**